### ISSUE TITLE
Light Shadows design issue?

### ISSUE THREAD

**Issue by neeh** [mrdoob/three.js#11035]
##### `WebGLRenderer` allocates one shadow map per light even for lights that are not supposed to cast shadows

This issue is related to [this question](http://stackoverflow.com/questions/42877678/many-lights-with-shadows-in-three-js-causes-fragment-shader-error/42889375#42889375).

As you can see on [shadowmap_pars_fragment.glsl](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl), a new shadow map is allocated on GPU for every light, even for those having `castShadow = false;`

    #ifdef USE_SHADOWMAP

        ...

        #if NUM_DIR_LIGHTS > 0

            // Reserving NUM_POINT_LIGHTS texture units
            uniform sampler2D directionalShadowMap[ NUM_POINT_LIGHTS ];
            varying vec4 vDirectionalShadowCoord[ NUM_POINT_LIGHTS ];

        #endif

        ...

    #endif

This is a problem because the GPU has a rather limited capacity of concurrent texture slots. The following error can occur when there are too much lights with `renderer.shadowMap.enabled = true` : 

> gl.getProgramInfoLog Fragment shader sampler count exceeds MAX_TEXTURE_IMAGE_UNITS (16).

On [this fiddle](https://jsfiddle.net/ro8j7z4z/3/), we're only using 4 concurrent shadow maps by moving some `LightShadow` and setting `light.castShadow = true` to the 4 nearest lights.
But if you try to increase the number of lights (just change `numLightRows`), You may see the error.

Additionally, You may also see the error : 

> Varyings over maximum register limit

Which is bound to the number of shadows coordinates that are sent to the fragment shader.

I think we should go back to something like [`allocateShadows()`](https://github.com/mrdoob/three.js/blob/8fcb3c06165d4616ba960fbfbb9e208fead9d64c/src/renderers/WebGLRenderer.js#L6059).
The shadow maps for `PointLight`s are sent [here](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L2419).

##### Three.js version

- [x] Dev
- [x] r84

##### Browser

- [x] All of them

##### OS

- [x] All of them

##### Hardware Requirements (graphics card, VR Device, ...)

Depends on maximum texture units.
Check the _Max Texture Image Units_ [here](https://browserleaks.com/webgl)
Or `new THREE.WebGLRenderer().capabilities.maxTextures;`

**Comment by cl4ws0n** [comment]
@neeh I just ran into this issue myself. Relieved to at least find I'm not the only one... Have you been able to come up with a temporary solution or anything? I'm pretty desperate as this is totally stopping my progress right now.

The fiddle doesn't work for me at all fwiw (`Uncaught TypeError: Cannot read property 'matrix' of null
    at setupLights (threejs.org/build/three.js:22287)`)

**Comment by arcs-** [comment]
@cl4ws0n I'm the guy from the stack overflow question, and I'm sorry, I've tried a lot but haven't found any solution. Eventually, I had to limit the number of lights. 
Indeed, the fiddle seems broken.

**Comment by neeh** [comment]
@cl4ws0n I just updated the fiddle for r85.
Unfortunately, The fiddle won't help you because the issue is inside Three.js and I wasn't able to hack around it.

After this fiddle, I worked on [another solution](https://jsfiddle.net/banjfscx/) that deals with 1 layer of **prelit textures**, 1 layer of `ShadowMaterial` and a fixed number of lights that are moved on the scene based on the actor's position. This can potentially lead to an infinite number of lights.
I don't remember why but this is broken. Firstly, I think there's a bug with `ShadowMaterial` that can't handle more than 1 light.

However, I'm pretty sure we can make the first fiddle work by overriding the `MeshPhongMaterial` and passing `LightShadows` as uniforms.

**Comment by cl4ws0n** [comment]
Hm. Interesting. I'd love it if you could elaborate on overriding the `MeshPongMaterial`. I'm definitely more interested in modifying the source to work as expected... by changing the standard shader fragments, either to rewrite the entire procedure that produces this problem (probably over my head) or just to add a `LODPointLight` and use a system like moving the shadow camera, but I don't really know what that'd entail either....I'll keep looking into it in the mean time.

**Comment by cl4ws0n** [comment]
@neeh Okay so I've spent some time digging into the source and it looks like in order to add a new type of light:

- Add a new `Light` type file at [`lights/`](https://github.com/mrdoob/three.js/tree/r85/src/lights)
- `export` it from [`Three`](https://github.com/mrdoob/three.js/blob/r85/src/Three.js#L52)
- In `WebGLRenderer`:
  - [Register it in the `_lights` `Array` cache](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L184)
  - [Store its `Array` inside `Material`s in `initMaterial`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L1635)
  - [Add it to `markUniformsLightsNeedsUpdate`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L2277)
  - [Add a length var and test case for it to `setupLights`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L2439)
  - [Clean up the cached `Array`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L2494)
  - [Update the `Hash`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L2498)
- Add a `case` for it in [`WebGLLights`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/webgl/WebGLLights.js#L56)
- Add a `.replace` step for it in `replaceLightNums` in [`WebGLProgram`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/webgl/WebGLProgram.js#L148)
- In `WebGLPrograms`:
  - [Add an entry for it in `parameterNames`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/webgl/WebGLPrograms.js#L34)
  - [Translate its `Array` length into a `Parameter` in `.getParameters`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/webgl/WebGLPrograms.js#L180)
- Add an entry for it in [`UniformsLib`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/UniformsLib.js#L134)
- Add a `case` for it in the `ShaderChunk`s:
  - [`lights_pars`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/lights_pars.glsl#L42)
  - [`lights_template`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/lights_template.glsl#L24)

And if the new light type has shadows:
- Store its `ShadowMap` and `ShadowMatrix` in `initMaterial` in [`WebGLRenderer`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/WebGLRenderer.js#L1642)
- Potentially modify the test case for `isPointLight` in [`WebGLShadowMap`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/webgl/WebGLShadowMap.js#L146)
- Add a `case` for it in the `ShaderChunk`s:
  - [`shadowmap_pars_fragment`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl#L17)
  - [`shadowmap_pars_vertex`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/shadowmap_pars_vertex.glsl#L17)
  - [`shadowmap_vertex`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/shadowmap_vertex.glsl#L23)
  - [`shadowmask_pars_fragment`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/shadowmask_pars_fragment.glsl#L33)

So clearly this is not the "easy option" I initially thought it was, just due to the number of modifications required. However, there is an additional problem I didn't realize to begin with. Shadows are directly tied to the lights, so even if we create another type of light that does not have ShadowMaps spammed for it, I am not really sure how to make a system utilizing them.

Anyway I will leave this here for now and keep working on it. Each link points to the line where the task I'm referring to is done for PointLights so this should at least make a good reference for further study

**Comment by cl4ws0n** [comment]
I am in the process of modifying the source to work as needed...So far I have the number of lights rendered clamped to a maximum working well. However, in trying to clamp shadows I have run into this problem, in [`lights_template.glsl`](https://github.com/mrdoob/three.js/blob/r85/src/renderers/shaders/ShaderChunk/lights_template.glsl#L34) :
```glsl
#if ( NUM_POINT_LIGHTS > 0 ) && defined( RE_Direct )

	PointLight pointLight;

	for ( int i = 0; i < NUM_POINT_LIGHTS; i ++ ) {

		pointLight = pointLights[ i ];

		getPointDirectLightIrradiance( pointLight, geometry, directLight );

		#ifdef USE_SHADOWMAP
		directLight.color *= all( bvec2( pointLight.shadow, directLight.visible ) ) ? getPointShadow( pointShadowMap[ i ], pointLight.shadowMapSize, pointLight.shadowBias, pointLight.shadowRadius, vPointShadowCoord[ i ] ) : 1.0;
		#endif

		RE_Direct( directLight, geometry, material, reflectedLight );

	}

#endif
```
Specifically `getPointShadow( pointShadowMap[ i ], pointLight.shadowMapSize, pointLight.shadowBias, pointLight.shadowRadius, vPointShadowCoord[ i ] )`

The `pointShadowMap` is not the same length as the `pointLights`, and may not be in the same order if a light near to the camera has `.visible = false`...

This should be trivial to fix for a more experienced shader programmer, but I haven't done much myself. My first thought was to add a uniform to the light, `pointLight.shadowIndex`, however it turns out array indices like this must be constant values. Trying to utilize loop unrolling the way the lights have here overloads the memory.

My next idea was to fully separate the shadows from the lights and iterate both independently. This is fairly trivial to do but as far as rewriting this particular file I'm still at a loss, because I can't simply iterate lights and then shadows as `getPointDirectLightIrradiance` overwrites the previous value of `directLight` and the shadow for that light needs to be applied in the loop step.

Any thoughts???

**Comment by cl4ws0n** [comment]
Alright @arcs- @neeh I finally have a working solution so I thought I'd share even though there are some limitations which prevent me from making a pull request/Stack answer. [Here's a diffy](https://diffy.org/diff/5eqkapvsh8f3i5l87prutvgqfr) / [raw diff](https://pastebin.com/xm7X0pZL), either of which you can apply to the source with `git apply thatfile.diff` for right now. The issues left are the following:
1. I don't know how to determine the maximum number of lights we can pass, so that is hard coded. It should really be an option for `WebGLRenderer` at the very least but I have left it until I can resolve whether its possible to make that determination.
2. All my stuff uses materials only, and no textures, so I'm not sure if `capabilities.maxTextures` is a safe option for the maximum number of shadows, so that is also hardcoded for now.

Other than that and some room for optimizations, it's all good afaik. Let me know if you run into any issues if you end up using it.

**Comment by cl4ws0n** [comment]
Can I just say, I think its really crazy that A.) this is an issue at all, that B.) there seems to be no interest at all in fixing it now, that C.) This is only labeled as a "suggestion" and finally D.) No one has pointed out how terrible my "solution" is or anything....Having tested it a bit I gotta say the performance penalty is huge. I'm just not qualified to make such a drastic change to the rendering code on my own at this point, though I will continue to work on it. It would be *great* if someone with some experience with this lib would chime in about my "solution" and offer some potential optimizations/design changes.

Until then though, I defy you to point out another graphics library that will *straight up crash* if you give it more than 15 lights and a single shadow. If this is the *intended* behavior, it should at least be documented and handled by the renderer *without crashing it.* 

TL;DR I insist this is a **bug**, and deserves more attention from experienced devs than it has gotten. Really *anything at all* would help me in the right direction at this point.

**Comment by darrelbelvin** [comment]
I agree with ghost! I'm working on a project where I need more than six lights. I could live with only a few of them casting shadows, but it's crashing even when I disable shadows for all but one light... **THIS IS A BUG!**

**Comment by richmerwarth** [comment]
I'm hitting this same problem here as well. 
Nothing helpful to add, I'm afraid, except another voice of frustration over this issue.

**Comment by darrelbelvin** [comment]
I came up with a solution for my application:

`var lightCount = x_div * z_div + 1,textureCount = textures.length,
	shadows = (lightCount <= renderer.capabilities.maxVaryings - 3) &&
			(lightCount + textureCount <= renderer.capabilities.maxTextures);
renderer.shadowMap.enabled = shadows;`

**Comment by mrdoob** [comment]
Fixed in https://github.com/mrdoob/three.js/pull/17074
